core/src/main/java/org/apache/calcite/rel/rules/JoinToCorrelateRule.java

```diff
@@ -127,7 +142,7 @@ public void onMatch(RelOptRuleCall call) {
             relBuilder.build(),
             correlationId,
             requiredColumns.build(),
-            SemiJoinType.of(join.getJoinType()));
+            join instanceof SemiJoin ? SemiJoinType.SEMI : SemiJoinType.of(join.getJoinType()));
```

Comment:
Maybe we could avoid this cast by introducing a more general method in Join which can bring back all the necessary types (SEMI, ANTI, etc.).

Reply:
What about a semiJoinTypeFunction lambda? New version committed with this approach

Reply:
I didn't have a function in mind but it can do the job. 

@julianhyde, I've seen that you had some concerns against a lamda function in the Jira issue so I am waiting for your input. From my perspective, if we keep this function and its configuration private I don't see a real problem.